Add cryptographic hash example

tarruda opened the pull request:
cc @agup006

tchrono committed:
Add cryptographic hash example

patrick-stephens commented on crypto-hash-example/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM fluent/fluent-bit:1.8-debug
```
1.9 surely?

patrick-stephens commented on crypto-hash-example/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+      luarocks && \
+    luarocks install luaossl
+
+RUN mkdir -pv /source
```
Not required as `WORKDIR` does it.

patrick-stephens commented on crypto-hash-example/Makefile:
```diff
@@ -0,0 +1,7 @@
+run-example: build
+	docker run --rm -ti --name example -v$(PWD):/source example /fluent-bit/bin/fluent-bit -c flb.conf
```
This was fail without a TTY (e.g. in CI). Remove the `i`.
It also assumes the current directory is the makefile directory, technically you could run make from another directory and it may fail to find the files.

patrick-stephens commented on crypto-hash-example/Makefile:
```diff
@@ -0,0 +1,7 @@
+	docker run --rm -ti --name example -v$(PWD):/source example /fluent-bit/bin/fluent-bit -c flb.conf
+
+build:
+	docker build -t example .
```
I'd go with a more descriptive image name...

patrick-stephens commented on crypto-hash-example/filter.lua:
```diff
@@ -0,0 +1,27 @@
+local digest = require 'openssl.digest'
```
Probably want a comment to explain what things do, pre/post conditions, parameters, etc. plus a copyright header.

patrick-stephens commented on crypto-hash-example/flb.conf:
```diff
@@ -0,0 +1,20 @@
+[INPUT]
+    name tail
+    read_from_head true
+    exit_on_eof true
```
I'd add a comment indicating this may not flush all output in case people see occasional dropped output.

patrick-stephens commented on crypto-hash-example/flb.conf:
```diff
@@ -0,0 +1,20 @@
+[FILTER]
+    Name    lua
+    Match   *
+    script  filter.lua
```
We could also have this all as inline LUA in YAML with 1.9...

tarruda replied:
Sure, we can do in a future iteration, too many changes in this PR already :D

patrick-stephens reviewed (commented):
Few niggles but looks good to me.
I'd suggest we add some CI for linting to pick up the niggles: hadolint, copyright headers, any LUA linters?

patrick-stephens commented:
My main suggestion would be we base it on the latest production image rather than a debug one. 
This requires a little work to install the extra dependencies but we could probably have a template Dockerfile to do it for all examples easily:
1. Separate stage to download required dependencies: https://github.com/fluent/fluent-bit/blob/f52b8ff234018912b7b02b9986d09fcd35f91631/dockerfiles/Dockerfile#L93
2. Pull the downloads into the main image on top of the production one: https://github.com/fluent/fluent-bit/blob/f52b8ff234018912b7b02b9986d09fcd35f91631/dockerfiles/Dockerfile#L162

Luarocks install is probably the thing I don't know here - what exactly does it do? Can we get it to run in a download-only mode? Can we copy anything it installs over? Do we actually need `luarocks` on the final image or just what it installs?

tarruda commented:
>Luarocks install is probably the thing I don't know here - what exactly does it do? Can we get it to run in a download-only mode? Can we copy anything it installs over? Do we actually need luarocks on the final image or just what it installs?

Luarocks can be uninstalled safely, all it does it build native code and install all modules in places where lua interpreter can find it.

patrick-stephens commented:
> > Luarocks install is probably the thing I don't know here - what exactly does it do? Can we get it to run in a download-only mode? Can we copy anything it installs over? Do we actually need luarocks on the final image or just what it installs?
> 
> Luarocks can be uninstalled safely, all it does it build native code and install all modules in places where lua interpreter can find it.

Or can we just build luaossl itself and copy the binaries?

tarruda commented:
> Or can we just build luaossl itself and copy the binaries?

Each lua package has its own build recipe that is followed by luarocks. Since there are many of these lua packages we may end up needing, isn't it more practical to delegate to luarocks to build/install, then uninstall it?

patrick-stephens commented:
> Each lua package has its own build recipe that is followed by luarocks. Since there are many of these lua packages we may end up needing, isn't it more practical to delegate to luarocks to build/install, then uninstall it?

Sure, although we would not do it that way but with a separate stage - there's no need to uninstall it, you just only copy the lua stuff over. However it depends on how consistent everything is - do lua modules get installed in a common location, etc.?

```
FROM debian:bullseye-slim as lua-installer
RUN apt-get update && apt-get install luarocks && luarocks install ....

FROM fluent/fluent-bit as production
COPY --from=lua-installer <lua-path> <lua-path>
```

tchrono committed:
Completely refactor Dockerfile/Makefile

- Now only one Dockerfile and one Makefile is used for all examples
- Each example specific Makefile/Dockerfile/README.md was removed
- Fluent-bit production image is used, with a separate step to build lua
  dependencies.

tchrono committed:
Add license and some comments on crypto hash example

tchrono committed:
Add -t flag to docker run so we get colored output

tarruda commented:
@patrick-stephens Now using a single Dockerfile/Makefile for all examples, also addressed other requests

tarruda commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    luarocks install --tree /luarocks xmlua && \
+    luarocks install --tree /luarocks base64
+
+FROM fluent/fluent-bit:1.9-debug
```
@patrick-stephens if I change this to `fluent/fluent-bit:1.9` things stop working. Any clues why?

patrick-stephens replied:
How so? I can't see any inherent reason it should. Sometimes the shell is used for commands but I think COPY should be fine here.

`fluent/fluent-bit:1.9` only recently got done for 1.9.4+ so it might not have been working prior to that as the tag was not present.

patrick-stephens replied:
It seems to build and run for me fine, I'm not using LUA stuff so do you mean something there does not work?

My guess is there are some extra dependencies we need to add that are not just the luarocks ones.

tarruda replied:
your'e right, I need to copy the native libs too

patrick-stephens replied:
Ah yeah, you need to split that out like we do here: https://github.com/fluent/fluent-bit/blob/0c41d2bf1a63a1b83a0323b2c87024537cbaad59/dockerfiles/Dockerfile#L101-L139

tarruda replied:
Done, I used a different method to extract the libraries, but hopefully this will be unnecessary in the future if we ship a standard library for lua filters.

tchrono committed:
Move flb.conf to root directory

tchrono committed:
Switch to using fluent-bit production image

patrick-stephens commented on Dockerfile:
```diff
@@ -0,0 +1,40 @@
+      | tar cf - -C / -h --verbatim-files-from -T - \
+      # Untar in the /luarocks directory, which is what we will unpack in the final image
+      | tar xf - -C /luarocks && \
+    # Same as above, but use dpkg -L to retrieve libraries not visible in ldd output. These are loaded through luajit FFI by xmlua
```
Lol super fun, so in production dockerfile we'll have to consolidate this down to the single approach I think that is already used but fine for now.

tarruda replied:
Yea, unfortunately xmlua is an exception because it relies on LuaJIT FFI to load libxml. Usually native Lua modules have an explicit link to system dependencies which would be visible in the `ldd` output.

In any case, I also think it is better to explicitly list dependencies, but I would like some extra time to write more Lua examples which will certainly require more dependencies. After we have a good set of Lua examples doing various tasks, we can define a standard library that would be included in the official image.
